src/components/atoms/Toggler/Toggler.tsx

```diff
@@ -0,0 +1,47 @@
+import classNames from "classnames";
+import React, {
+  DetailedHTMLProps,
+  forwardRef,
+  InputHTMLAttributes,
+} from "react";
+import { v4 as uuidv4 } from "uuid";
+
+import "./Toggler.scss";
+
+export interface TogglerProps
+  extends DetailedHTMLProps<
+    InputHTMLAttributes<HTMLInputElement>,
+    HTMLInputElement
+  > {
+  title?: string;
+  containerClassName?: string;
+  inputClassName?: string;
+}
+
+const Toggler: React.ForwardRefRenderFunction<HTMLDivElement, TogglerProps> = (
+  {
+    containerClassName,
+    inputClassName,
+    title,
+    // In case id is not provided, generate one. It's requires so that htmlFor works correctly
+    id = uuidv4(),
+    ...extraInputProps
+  },
+  ref
+) => {
+  const sliderClasses = classNames("Toggler__slider", {
+    "Toggler__slider--checked": extraInputProps.checked,
+  });
+
+  return (
+    <div ref={ref} className="Toggler">
+      <label className="Toggler__input" htmlFor={id}>
+        <span className={sliderClasses} />
+      </label>
+      <input hidden type="checkbox" {...extraInputProps} id={id} />
+      {title && <div className="Toggler__title">{title}</div>}
+    </div>
+  );
+};
+
+export default forwardRef(Toggler);
```

Comment:
That's becase of the forwardRef. index.ts of this component exports it as a named one

Reply:
Can’t you just in-line forwardRef into the component definition, or create a second named const to export. I believe we already do this in other places in the app.

Reply:
Reworked in https://github.com/sparkletown/sparkle/pull/1299/commits/13d2c8b1bc477d69ee9814144c4fd09d3e1383a6

Reply:
> Can’t you just in-line forwardRef into the component definition

I expected this to look something like this:

```
export const Toggler: React.FC<TogglerProps> = forwardRef({
```

Yet your changes in https://github.com/sparkletown/sparkle/commit/13d2c8b1bc477d69ee9814144c4fd09d3e1383a6 don't seem to include the call to `forwardRef` anywhere at all anymore, instead creating a normal prop with the same name?

Reply:
It has some type errors, which I have no mental energy to resolve

![image](https://user-images.githubusercontent.com/63194656/117043857-6b89a280-ad16-11eb-9fb1-9c8ecee778b5.png)

Reply:
It's hard to give specific answers since I can't actually see the changes you tried making in your code in that screenshot.

---

Looking at:

- https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/

This should be a minimal/fully worked example:

```typescript
interface FooProps {
  foo: string;
}

type FooRefType = HTMLDivElement; // as an example

const foo: React.FC<FooProps> = forwardRef<FooRefType, FooProps>(
  (props, ref) => {
    return <div ref={ref} />;
  }
);
```

My guess is that because you didn't explicitly pass the types to `forwardRef` it ended up using something like `unknown` or similar for them, which was then incompatible with the `React.FC<TogglerProps>` type you had set.

It might also just be an issue with how you're defining the type for the ref.

Hard to say without further information/context/seeing the actual code you used.